From 62c6a69f4b9e5952860e48db783109b6481a3da5 Mon Sep 17 00:00:00 2001
From: Zhi Wang <zhi.a.wang@intel.com>
Date: Tue, 10 Feb 2015 23:24:09 +0800
Subject: [PATCH 15/18] xen: introduce unlimited rangeset.

Signed-off-by: Zhi Wang <zhi.a.wang@intel.com>
diff --git a/xen/common/rangeset.c b/xen/common/rangeset.c
index 6c6293c..b12be81 100644
--- a/xen/common/rangeset.c
+++ b/xen/common/rangeset.c
@@ -98,7 +98,7 @@ static struct range *alloc_range(
 {
     struct range *x;
 
-    if ( r->nr_ranges == 0 )
+    if ( !(r->flags & RANGESETF_unlimited) && r->nr_ranges == 0 )
         return NULL;
 
     x = xmalloc(struct range);
@@ -335,7 +335,7 @@ struct rangeset *rangeset_new(
     INIT_LIST_HEAD(&r->range_list);
     r->nr_ranges = -1;
 
-    BUG_ON(flags & ~RANGESETF_prettyprint_hex);
+    BUG_ON(flags & ~(RANGESETF_prettyprint_hex | RANGESETF_unlimited));
     r->flags = flags;
 
     if ( name != NULL )
diff --git a/xen/include/xen/rangeset.h b/xen/include/xen/rangeset.h
index aa64082..caf08d0 100644
--- a/xen/include/xen/rangeset.h
+++ b/xen/include/xen/rangeset.h
@@ -51,6 +51,8 @@ void rangeset_limit(
  /* Pretty-print range limits in hexadecimal. */
 #define _RANGESETF_prettyprint_hex 0
 #define RANGESETF_prettyprint_hex  (1U << _RANGESETF_prettyprint_hex)
+#define _RANGESETF_unlimited 1
+#define RANGESETF_unlimited (1U << _RANGESETF_unlimited)
 
 bool_t __must_check rangeset_is_empty(
     const struct rangeset *r);
