diff --git a/mk/xen.spec.in b/mk/xen.spec.in
index f9a79b5..416207d 100644
--- a/mk/xen.spec.in
+++ b/mk/xen.spec.in
@@ -321,6 +321,7 @@ rm -rf $RPM_BUILD_ROOT
 %{_includedir}/%{name}/physdev.h
 %{_includedir}/%{name}/platform.h
 %{_includedir}/%{name}/pmu.h
+%{_includedir}/%{name}/pv-iommu.h
 %{_includedir}/%{name}/sched.h
 %{_includedir}/%{name}/sys/evtchn.h
 %{_includedir}/%{name}/sys/gntalloc.h
