diff --git a/tools/config.h.in b/tools/config.h.in
index 478a2cc..b3fdfae 100644
--- a/tools/config.h.in
+++ b/tools/config.h.in
@@ -93,6 +93,9 @@
 /* Define extfs header to use */
 #undef INCLUDE_EXTFS_H
 
+/* Define json header to use */
+#undef INCLUDE_JSON_H
+
 /* libutil header file name */
 #undef INCLUDE_LIBUTIL_H
 
diff --git a/tools/configure b/tools/configure
index 2260ea4..98eea94 100755
--- a/tools/configure
+++ b/tools/configure
@@ -8849,6 +8849,32 @@ fi
 
 
 
+ac_fn_c_check_header_mongrel "$LINENO" "json/json.h" "ac_cv_header_json_json_h" "$ac_includes_default"
+if test "x$ac_cv_header_json_json_h" = xyes; then :
+
+
+$as_echo "#define INCLUDE_JSON_H <json/json.h>" >>confdefs.h
+
+
+else
+
+    ac_fn_c_check_header_mongrel "$LINENO" "json-c/json.h" "ac_cv_header_json_c_json_h" "$ac_includes_default"
+if test "x$ac_cv_header_json_c_json_h" = xyes; then :
+
+
+$as_echo "#define INCLUDE_JSON_H <json-c/json.h>" >>confdefs.h
+
+
+else
+  as_fn_error $? "Could not find libjson" "$LINENO" 5
+fi
+
+
+
+fi
+
+
+
 # FDT is needed only on ARM
 case "$host_cpu" in
 arm*|aarch64)
diff --git a/tools/configure.ac b/tools/configure.ac
index f164694..328d835 100644
--- a/tools/configure.ac
+++ b/tools/configure.ac
@@ -374,6 +374,14 @@ AC_CHECK_LIB([argp], [argp_usage], [argp_ldflags="-largp"])
 ], [AC_MSG_ERROR([Could not find argp])])
 AC_SUBST(argp_ldflags)
 
+AC_CHECK_HEADER([json/json.h], [
+    AC_DEFINE([INCLUDE_JSON_H], [<json/json.h>], [Define json header to use])
+], [
+    AC_CHECK_HEADER([json-c/json.h], [
+        AC_DEFINE([INCLUDE_JSON_H], [<json-c/json.h>], [Define json header to use])
+    ], [AC_MSG_ERROR([Could not find libjson])])
+])
+
 # FDT is needed only on ARM
 case "$host_cpu" in
 arm*|aarch64)
