From 668ba1f85bf2e4086cf18c35abc880b9eee4e8f2 Mon Sep 17 00:00:00 2001
From: Daniel De Graaf <dgdegra@tycho.nsa.gov>
Date: Mon, 20 Jun 2016 10:04:25 -0400
Subject: [PATCH] xen: Make FLASK_AVC_STATS kconfig option visible

Signed-off-by: Daniel De Graaf <dgdegra@tycho.nsa.gov>
Reviewed-by: Doug Goldstein <cardoe@cardoe.com>
diff --git a/xen/common/Kconfig b/xen/common/Kconfig
index faee3ec..daab832 100644
--- a/xen/common/Kconfig
+++ b/xen/common/Kconfig
@@ -134,9 +134,14 @@ config FLASK
 
 config FLASK_AVC_STATS
 	def_bool y
+	prompt "Maintain statistics on the FLASK access vector cache" if EXPERT = "y"
 	depends on FLASK
 	---help---
-	  Maintain statistics on the access vector cache
+	  Maintain counters on the access vector cache that can be viewed using
+	  the FLASK_AVC_CACHESTATS sub-op of the xsm_op hypercall.  Disabling
+	  this will save a tiny amount of memory and time to update the stats.
+
+	  If unsure, say Y.
 
 # Enable schedulers
 menu "Schedulers"
