diff --git a/.gitattributes b/.gitattributes
new file mode 100644
index 0000000..7e70d4c
--- /dev/null
+++ b/.gitattributes
@@ -0,0 +1,9 @@
+.bk-to-hg export-ignore
+.hg-to-bk export-ignore
+.hgignore export-ignore
+.hgsigs export-ignore
+.hgtags export-ignore
+.rootkeys export-ignore
+.gitignore export-ignore
+.gitattributes export-ignore
+mk/ export-ignore
diff --git a/mk/Makefile b/mk/Makefile
new file mode 100644
index 0000000..96c395d
--- /dev/null
+++ b/mk/Makefile
@@ -0,0 +1,128 @@
+USE_BRANDING := yes
+IMPORT_BRANDING := yes
+# makefile for xen component in build system
+include $(B_BASE)/common.mk
+include $(B_BASE)/rpmbuild.mk
+
+# SHELL := /bin/bash
+# SHELLOPTS := errexit:pipefail
+# export SHELL SHELLOPTS
+
+quiet ?= 0
+ifneq ($(quiet),0)
+  Q := @
+else
+  Q :=
+endif
+
+.PHONY: %var
+%var:
+	@echo "$* = $($*)"
+
+# Pull repo name from $PWD rather than hard coding it
+REPO_NAME := $(shell echo "$$PWD" | awk -F/ '{print $$3}')
+RPM_BUILD_COOKIE := $(MY_OBJ_DIR)/.rpm_build_cookie
+REPO_STAMP := $(call git_req,$(REPO_NAME))
+
+CDFILES_COOKIE = $(MY_OBJ_DIR)/.cdfiles.cookie
+
+$(eval $(shell $(call git_cset_number,$(REPO_NAME)))) # Defines CSET_NUMBER for us
+XEN_VERSION := $(shell $(MAKE) -C $(call git_loc,$(REPO_NAME)) --no-print-directory xenversion)
+XEN_RELEASE := $(PLATFORM_VERSION).$(CSET_NUMBER)
+
+ifeq ($(BUILD_NUMBER),0x)
+XEN_VENDORVERSION := -xs-local
+else
+XEN_VENDORVERSION := -xs$(shell echo $(BUILD_NUMBER) | tr -dc [:digit:])
+endif
+
+REPO_LOC := $(call git_loc,$(REPO_NAME))
+BASE_CSET_STR := $(shell { git rev-parse --verify --short=12 qparent --git-dir=$(REPO_LOC)/.git; } )
+PQ_CSET_STR   := $(shell { git rev-parse --verify --short=12 HEAD    --git-dir=$(REPO_LOC)/.git/patches/.git; } )
+
+XEN_CHANGESET := "$(BASE_CSET_STR), pq $(PQ_CSET_STR)"
+
+.PHONY: build
+build: $(RPM_BUILD_COOKIE) $(BRANDING) $(MY_OUTPUT_DIR)/xen.inc \
+       $(CDFILES_COOKIE)
+	@ :
+
+# # Xen packages
+
+SOURCES := $(RPM_SOURCESDIR)/xen-$(XEN_VERSION).tar.bz2
+SOURCES += $(RPM_SOURCESDIR)/sysconfig_kernel-xen
+SOURCES += $(RPM_SOURCESDIR)/xl.conf
+SOURCES += $(RPM_SOURCESDIR)/logrotate-xen-tools
+SOURCES += $(RPM_SPECSDIR)/xen.spec
+SOURCES += $(RPM_SOURCESDIR)/xen-development.patch
+
+XEN_HG_EXCLUDE := -X mk
+
+$(RPM_SOURCESDIR)/xen-$(XEN_VERSION).tar.bz2: $(RPM_SOURCESDIR)/.dirstamp
+	{ set -e; set -o pipefail; \
+	cd $(REPO_LOC); \
+	git archive --format=tar --prefix=xen-$(XEN_VERSION)/ HEAD | bzip2 -9 > $@.tmp; \
+	mv -f $@.tmp $@; \
+	}
+
+$(RPM_SOURCESDIR)/%: %
+	cp $^ $@
+
+$(RPM_SOURCESDIR)/xen-development.patch: $(RPM_SOURCESDIR)/.dirstamp
+	{ set -e; set -o pipefail; \
+	cd $(REPO_LOC); \
+	git diff HEAD > $@.tmp; \
+	mv -f $@.tmp $@; \
+	}
+
+$(RPM_SPECSDIR)/xen.spec: xen.spec.in $(RPM_SPECSDIR)/.dirstamp Makefile
+	{ set -e; set -o pipefail; \
+	sed -e s/@XEN_VERSION@/$(XEN_VERSION)/g \
+	    -e s/@XEN_RELEASE@/$(XEN_RELEASE)/g \
+	    -e s/@XEN_CHANGESET@/$(XEN_CHANGESET)/g \
+	    -e s/@XEN_VENDORVERSION@/$(XEN_VENDORVERSION)/g \
+	< $< > $@.tmp; \
+	mv -f $@.tmp $@; \
+	}
+
+$(RPM_BUILD_COOKIE): $(RPM_DIRECTORIES) $(SOURCES)
+	$(RPMBUILD) --target $(DOMAIN0_ARCH_OPTIMIZED) -ba $(RPM_SPECSDIR)/xen.spec
+	$(Q)touch $@
+
+# # Files for CD
+HYPERVISOR_OUTPUT_RPM := xen-hypervisor-$(XEN_VERSION)-$(XEN_RELEASE).$(DOMAIN0_ARCH_OPTIMIZED).rpm
+
+$(CDFILES_COOKIE): $(RPM_BUILD_COOKIE) $(MY_MAIN_CDFILES)/.dirstamp
+	rpm2cpio $(MY_OUTPUT_DIR)/RPMS/$(DOMAIN0_ARCH_OPTIMIZED)/$(HYPERVISOR_OUTPUT_RPM) | \
+		( cd $(MY_MAIN_CDFILES) ; cpio --extract --make-directories --unconditional ./boot/xen-$(XEN_VERSION)$(XEN_VENDORVERSION)-d.gz )
+	mv $(MY_MAIN_CDFILES)/boot/xen-$(XEN_VERSION)$(XEN_VENDORVERSION)-d.gz $(MY_MAIN_CDFILES)/boot/xen.gz
+	touch $@
+
+# # SOURCES, BRANDING etc
+
+.PHONY: $(MY_OUTPUT_DIR)/xen.inc
+$(MY_OUTPUT_DIR)/xen.inc: $(MY_OUTPUT_DIRSTAMP)
+	{ set -e; set -o pipefail; \
+	$(Q){ echo XEN_PKG_NAME := xen; \
+	  echo XEN_PKG_VERSION := $(XEN_VERSION)-$(XEN_RELEASE); \
+	  echo XEN_PKG_ARCH := $(DOMAIN0_ARCH_OPTIMIZED); \
+	  echo XEN_PKG_FILE := RPMS/$(DOMAIN0_ARCH_OPTIMIZED)/$(OUTPUT_RPM); \
+	  echo XEN_VERSION := $(XEN_VERSION); \
+	  echo XEN_RELEASE := $(XEN_RELEASE); \
+	} > $@.tmp; \
+	mv -f $@.tmp $@; \
+	}
+
+.PHONY: $(BRANDING)
+$(BRANDING): $(MY_OUTPUT_DIR)/.dirstamp
+	{ set -e; set -o pipefail; \
+	$(Q)echo "XEN_VERSION := $(XEN_VERSION)" > $@.tmp; \
+	mv -f $@.tmp $@; \
+	}
+
+.PHONY: clean
+clean:
+	rm -f $(RPM_BUILD_COOKIE)
+	rm -f $(SOURCES)
+	rm -f $(SOURCES:%=%.tmp)
+	rm -f $(MY_OUTPUT_DIR)/xen.inc
diff --git a/mk/logrotate-xen-tools b/mk/logrotate-xen-tools
new file mode 100644
index 0000000..2c84dc7
--- /dev/null
+++ b/mk/logrotate-xen-tools
@@ -0,0 +1,13 @@
+/var/log/xen/*.log {
+    missingok
+    postrotate
+        /usr/bin/systemctl kill -s SIGHUP xenconsoled.service
+    endscript
+    nocreate
+}
+
+/var/log/xenstored-access.log {
+    postrotate
+        /bin/kill -HUP `cat /var/run/syslogd.pid 2> /dev/null` 2> /dev/null || true
+    endscript
+}
diff --git a/mk/sysconfig_kernel-xen b/mk/sysconfig_kernel-xen
new file mode 100644
index 0000000..660b280
--- /dev/null
+++ b/mk/sysconfig_kernel-xen
@@ -0,0 +1,6 @@
+
+# Xen hypervisor
+HYPERVISOR=/boot/xen.gz
+# host-specific configuration belongs in /etc/sysconfig/kernel-xen-args
+[ -e /etc/sysconfig/kernel-xen-args ] && . /etc/sysconfig/kernel-xen-args
+HYPERVISOR_ARGS="$HYPERVISOR_ARGS watchdog"
diff --git a/mk/xen.spec.in b/mk/xen.spec.in
new file mode 100644
index 0000000..6a25659
--- /dev/null
+++ b/mk/xen.spec.in
@@ -0,0 +1,741 @@
+# -*- rpm-spec -*-
+
+%{!?python_sitearch: %define python_sitearch %(%{__python} -c "from distutils.sysconfig import get_python_lib; print get_python_lib(1)")}
+
+%define with_sysv 0
+%define with_systemd 1
+
+%define COMMON_OPTIONS DESTDIR=%{buildroot} %{?_smp_mflags}
+
+# For 32bit dom0 userspace, we need to cross compile a 64bit Xen
+%ifarch %ix86
+%define HVSOR_OPTIONS %{COMMON_OPTIONS} XEN_TARGET_ARCH=x86_64 CROSS_COMPILE=x86_64-linux-gnu-
+%define TOOLS_OPTIONS %{COMMON_OPTIONS} XEN_TARGET_ARCH=x86_32 debug=n
+%endif
+
+%define vendor_version @XEN_VENDORVERSION@
+
+# For 64bit
+%ifarch x86_64
+%define HVSOR_OPTIONS %{COMMON_OPTIONS} XEN_TARGET_ARCH=x86_64
+%define TOOLS_OPTIONS %{COMMON_OPTIONS} XEN_TARGET_ARCH=x86_64 debug=n
+%endif
+
+Summary: Xen is a virtual machine monitor
+Name:    xen
+Version: @XEN_VERSION@
+Release: @XEN_RELEASE@
+License: GPL
+URL:     http://www.xen.org
+Source0: %{name}-%{version}.tar.bz2
+Source1: sysconfig_kernel-xen
+Source2: xl.conf
+Source3: logrotate-xen-tools
+Patch0:  xen-development.patch
+
+ExclusiveArch: i686 x86_64
+
+#Cross complier
+%ifarch %ix86
+BuildRequires: gcc-x86_64-linux-gnu binutils-x86_64-linux-gnu
+%endif
+
+# For HVMLoader and 16/32bit firmware
+BuildRequires: /usr/include/gnu/stubs-32.h
+BuildRequires: dev86 iasl
+
+# For the domain builder (decompression and hashing)
+BuildRequires: zlib-devel bzip2-devel xz-devel
+BuildRequires: openssl-devel
+
+# For libxl
+BuildRequires: yajl-devel libuuid-devel perl
+
+# For python stubs
+BuildRequires: python-devel
+
+# For ocaml stubs
+BuildRequires: ocaml ocaml-findlib
+
+# For ipxe
+BuildRequires: ipxe-source
+
+BuildRequires: libblkid-devel
+
+# For xentop
+BuildRequires: ncurses-devel
+
+# For libfsimage
+BuildRequires: e2fsprogs-devel
+%if 0%{?centos}%{!?centos:5} < 6 && 0%{?rhel}%{!?rhel:5} < 6
+#libext4fs
+BuildRequires: e4fsprogs-devel
+%endif
+
+# For xenguest
+BuildRequires: json-c-devel
+
+# Misc
+BuildRequires: libtool
+%if %with_systemd
+BuildRequires: systemd-devel
+%endif
+
+# To placate ./configure
+BuildRequires: gettext-devel glib2-devel curl-devel gnutls-devel
+
+%description
+Xen Hypervisor.
+
+%package hypervisor
+Summary: The Xen Hypervisor
+Group: System/Hypervisor
+Requires(post): coreutils grep
+%description hypervisor
+This package contains the Xen Hypervisor.
+
+%package hypervisor-debuginfo
+Summary: The Xen Hypervisor debug information
+Group: Development/Debug
+%description hypervisor-debuginfo
+This package contains the Xen Hypervisor debug information.
+
+%package tools
+Summary: Xen Hypervisor general tools
+Requires: xen-libs = %{version}
+Group: System/Base
+%description tools
+This package contains the Xen Hypervisor general tools for all domains.
+
+%package devel
+Summary: The Xen Hypervisor public headers
+Group: Development/Libraries
+%description devel
+This package contains the Xen Hypervisor public header files.
+
+%package libs
+Summary: Xen Hypervisor general libraries
+Group: System/Libraries
+%description libs
+This package contains the Xen Hypervisor general libraries for all domains.
+
+%package libs-devel
+Summary: Xen Hypervisor general development libraries
+Requires: xen-libs = %{version}
+Requires: xen-devel = %{version}
+Group: Development/Libraries
+%description libs-devel
+This package contains the Xen Hypervisor general development for all domains.
+
+%package dom0-tools
+Summary: Xen Hypervisor Domain 0 tools
+Requires: xen-dom0-libs = %{version}
+Requires: xen-tools = %{version}
+%if %with_systemd
+Requires(post): systemd
+Requires(preun): systemd
+Requires(postun): systemd
+BuildRequires: systemd
+%endif
+Group: System/Base
+%description dom0-tools
+This package contains the Xen Hypervisor control domain tools.
+
+%package dom0-libs
+Summary: Xen Hypervisor Domain 0 libraries
+Requires: xen-hypervisor = %{version}
+Group: System/Libraries
+%description dom0-libs
+This package contains the Xen Hypervisor control domain libraries.
+
+%package dom0-libs-devel
+Summary: Xen Hypervisor Domain 0 headers
+Requires: xen-devel = %{version}
+Requires: xen-dom0-libs = %{version}
+
+# Temp until the build dependencies are properly propagated
+Provides: xen-dom0-devel = %{version}
+Group: Development/Libraries
+%description dom0-libs-devel
+This package contains the Xen Hypervisor control domain headers.
+
+%package ocaml-libs
+Summary: Xen Hypervisor ocaml libraries
+Requires: xen-dom0-libs = %{version}
+Group: System/Libraries
+%description ocaml-libs
+This package contains the Xen Hypervisor ocaml libraries.
+
+%package ocaml-devel
+Summary: Xen Hypervisor ocaml headers
+Requires: xen-ocaml-libs = %{version}
+Requires: xen-dom0-libs-devel = %{version}
+Group: Development/Libraries
+%description ocaml-devel
+This package contains the Xen Hypervisor ocaml headers.
+
+%package installer-files
+Summary: Xen files for the XenServer installer
+Group: System Environment/Base
+%description installer-files
+This package contains the minimal subset of libraries and binaries required in
+the XenServer installer environment.
+
+%prep
+%setup -q
+mkdir -p tools/firmware/etherboot/ipxe/
+cp /usr/src/ipxe-source.tar.gz tools/firmware/etherboot/ipxe.tar.gz
+rm -f tools/firmware/etherboot/patches/series
+%patch0 -p1 -b ~development
+echo "@XEN_CHANGESET@" > .scmversion
+
+%build
+
+# Placate ./configure, but don't pull in external content.
+export WGET=/bin/false FETCHER=/bin/false
+
+%configure \
+        --disable-seabios --disable-stubdom --disable-xsmpolicy --disable-blktap2 \
+	--with-system-qemu=%{_libdir}/xen/bin/qemu-system-i386 --with-xenstored=oxenstored \
+	--enable-systemd
+
+%install
+
+# The existence of this directory causes ocamlfind to put things in it
+mkdir -p %{buildroot}%{_libdir}/ocaml/stublibs
+
+mkdir -p %{buildroot}/boot/
+
+# Regular build of Xen
+%{__make} %{HVSOR_OPTIONS} -C xen XEN_VENDORVERSION=%{vendor_version} \
+    KCONFIG_CONFIG=../buildconfigs/config-release olddefconfig
+%{__make} %{HVSOR_OPTIONS} -C xen XEN_VENDORVERSION=%{vendor_version} \
+    KCONFIG_CONFIG=../buildconfigs/config-release build
+%{__make} %{HVSOR_OPTIONS} -C xen XEN_VENDORVERSION=%{vendor_version} \
+    KCONFIG_CONFIG=../buildconfigs/config-release MAP
+
+cp xen/xen.gz %{buildroot}/boot/%{name}-%{version}%{vendor_version}.gz
+cp xen/System.map %{buildroot}/boot/%{name}-%{version}%{vendor_version}.map
+cp xen/xen-syms %{buildroot}/boot/%{name}-syms-%{version}%{vendor_version}
+cp buildconfigs/config-release %{buildroot}/boot/%{name}-%{version}%{vendor_version}.config
+
+# Debug build of Xen
+%{__make} %{HVSOR_OPTIONS} -C xen clean
+%{__make} %{HVSOR_OPTIONS} -C xen XEN_VENDORVERSION=%{vendor_version}-d \
+    KCONFIG_CONFIG=../buildconfigs/config-debug olddefconfig
+%{?cov_wrap} %{__make} %{HVSOR_OPTIONS} -C xen XEN_VENDORVERSION=%{vendor_version}-d \
+    KCONFIG_CONFIG=../buildconfigs/config-debug build
+%{__make} %{HVSOR_OPTIONS} -C xen XEN_VENDORVERSION=%{vendor_version}-d \
+    KCONFIG_CONFIG=../buildconfigs/config-debug MAP
+
+cp xen/xen.gz %{buildroot}/boot/%{name}-%{version}%{vendor_version}-d.gz
+cp xen/System.map %{buildroot}/boot/%{name}-%{version}%{vendor_version}-d.map
+cp xen/xen-syms %{buildroot}/boot/%{name}-syms-%{version}%{vendor_version}-d
+cp buildconfigs/config-debug %{buildroot}/boot/%{name}-%{version}%{vendor_version}-d.config
+
+# do not strip the hypervisor-debuginfo targerts
+chmod -x %{buildroot}/boot/xen-syms-*
+
+# Build tools and man pages
+%{?cov_wrap} %{__make} %{TOOLS_OPTIONS} -C tools install
+%{__make} %{TOOLS_OPTIONS} -C docs install-man-pages
+%{?cov_wrap} %{__make} %{TOOLS_OPTIONS} -C tools/tests/mce-test/tools install
+
+%{__install} -D -m 644 %{SOURCE1} %{buildroot}%{_sysconfdir}/sysconfig/kernel-xen
+%{__install} -D -m 644 %{SOURCE2} %{buildroot}%{_sysconfdir}/xen/xl.conf
+%{__install} -D -m 644 %{SOURCE3} %{buildroot}%{_sysconfdir}/logrotate.d/xen-tools
+
+%files hypervisor
+/boot/%{name}-%{version}%{vendor_version}.gz
+/boot/%{name}-%{version}%{vendor_version}.map
+/boot/%{name}-%{version}%{vendor_version}.config
+/boot/%{name}-%{version}%{vendor_version}-d.gz
+/boot/%{name}-%{version}%{vendor_version}-d.map
+/boot/%{name}-%{version}%{vendor_version}-d.config
+%config %{_sysconfdir}/sysconfig/kernel-xen
+%ghost %attr(0644,root,root) %{_sysconfdir}/sysconfig/kernel-xen-args
+
+%files hypervisor-debuginfo
+/boot/%{name}-syms-%{version}%{vendor_version}
+/boot/%{name}-syms-%{version}%{vendor_version}-d
+
+%files tools
+%{_bindir}/xenstore
+%{_bindir}/xenstore-chmod
+%{_bindir}/xenstore-control
+%{_bindir}/xenstore-exists
+%{_bindir}/xenstore-list
+%{_bindir}/xenstore-ls
+%{_bindir}/xenstore-read
+%{_bindir}/xenstore-rm
+%{_bindir}/xenstore-watch
+%{_bindir}/xenstore-write
+%{python_sitearch}/%{name}/__init__.py*
+%{python_sitearch}/%{name}/lowlevel/__init__.py*
+%{python_sitearch}/%{name}/lowlevel/xs.so
+
+%files devel
+%{_includedir}/%{name}/COPYING
+%{_includedir}/%{name}/arch-arm.h
+%{_includedir}/%{name}/arch-arm/hvm/save.h
+%{_includedir}/%{name}/arch-x86/cpuid.h
+%{_includedir}/%{name}/arch-x86/cpufeatureset.h
+%{_includedir}/%{name}/arch-x86/hvm/save.h
+%{_includedir}/%{name}/arch-x86/pmu.h
+%{_includedir}/%{name}/arch-x86/xen-mca.h
+%{_includedir}/%{name}/arch-x86/xen-x86_32.h
+%{_includedir}/%{name}/arch-x86/xen-x86_64.h
+%{_includedir}/%{name}/arch-x86/xen.h
+%{_includedir}/%{name}/arch-x86_32.h
+%{_includedir}/%{name}/arch-x86_64.h
+%{_includedir}/%{name}/callback.h
+%{_includedir}/%{name}/dom0_ops.h
+%{_includedir}/%{name}/domctl.h
+%{_includedir}/%{name}/elfnote.h
+%{_includedir}/%{name}/errno.h
+%{_includedir}/%{name}/event_channel.h
+%{_includedir}/%{name}/features.h
+%{_includedir}/%{name}/foreign/arm32.h
+%{_includedir}/%{name}/foreign/arm64.h
+%{_includedir}/%{name}/foreign/x86_32.h
+%{_includedir}/%{name}/foreign/x86_64.h
+%{_includedir}/%{name}/grant_table.h
+%{_includedir}/%{name}/hvm/e820.h
+%{_includedir}/%{name}/hvm/hvm_info_table.h
+%{_includedir}/%{name}/hvm/hvm_op.h
+%{_includedir}/%{name}/hvm/hvm_vcpu.h
+%{_includedir}/%{name}/hvm/hvm_xs_strings.h
+%{_includedir}/%{name}/hvm/ioreq.h
+%{_includedir}/%{name}/hvm/params.h
+%{_includedir}/%{name}/hvm/pvdrivers.h
+%{_includedir}/%{name}/hvm/save.h
+%{_includedir}/%{name}/io/blkif.h
+%{_includedir}/%{name}/io/console.h
+%{_includedir}/%{name}/io/fbif.h
+%{_includedir}/%{name}/io/fsif.h
+%{_includedir}/%{name}/io/kbdif.h
+%{_includedir}/%{name}/io/libxenvchan.h
+%{_includedir}/%{name}/io/netif.h
+%{_includedir}/%{name}/io/pciif.h
+%{_includedir}/%{name}/io/protocols.h
+%{_includedir}/%{name}/io/ring.h
+%{_includedir}/%{name}/io/tpmif.h
+%{_includedir}/%{name}/io/usbif.h
+%{_includedir}/%{name}/io/vscsiif.h
+%{_includedir}/%{name}/io/xenbus.h
+%{_includedir}/%{name}/io/xs_wire.h
+%{_includedir}/%{name}/kexec.h
+%{_includedir}/%{name}/memory.h
+%{_includedir}/%{name}/nmi.h
+%{_includedir}/%{name}/physdev.h
+%{_includedir}/%{name}/platform.h
+%{_includedir}/%{name}/pmu.h
+%{_includedir}/%{name}/sched.h
+%{_includedir}/%{name}/sys/evtchn.h
+%{_includedir}/%{name}/sys/gntalloc.h
+%{_includedir}/%{name}/sys/gntdev.h
+%{_includedir}/%{name}/sys/privcmd.h
+%{_includedir}/%{name}/sys/xenbus_dev.h
+%{_includedir}/%{name}/sysctl.h
+%{_includedir}/%{name}/tmem.h
+%{_includedir}/%{name}/trace.h
+%{_includedir}/%{name}/vcpu.h
+%{_includedir}/%{name}/version.h
+%{_includedir}/%{name}/vm_event.h
+%{_includedir}/%{name}/xen-compat.h
+%{_includedir}/%{name}/xen.h
+%{_includedir}/%{name}/xencomm.h
+%{_includedir}/%{name}/xenoprof.h
+%{_includedir}/%{name}/xsm/flask_op.h
+
+%files libs
+%{_libdir}/libxenevtchn.so.1
+%{_libdir}/libxenevtchn.so.1.0
+%{_libdir}/libxengnttab.so.1
+%{_libdir}/libxengnttab.so.1.0
+%{_libdir}/libxenstore.so.3.0
+%{_libdir}/libxenstore.so.3.0.3
+%{_libdir}/libxenvchan.so.4.7
+%{_libdir}/libxenvchan.so.4.7.0
+
+%files libs-devel
+# Lib Xen Evtchn
+%{_includedir}/xenevtchn.h
+%{_libdir}/libxenevtchn.a
+%{_libdir}/libxenevtchn.so
+
+# Lib Xen Gnttab
+%{_includedir}/xengnttab.h
+%{_libdir}/libxengnttab.a
+%{_libdir}/libxengnttab.so
+
+# Lib XenStore
+%{_includedir}/xenstore.h
+%{_includedir}/xenstore_lib.h
+%{_libdir}/libxenstore.a
+%{_libdir}/libxenstore.so
+# Legacy XenStore header files, excluded to discourage their use
+%exclude %{_includedir}/xs.h
+%exclude %{_includedir}/xenstore-compat/xs.h
+%exclude %{_includedir}/xs_lib.h
+%exclude %{_includedir}/xenstore-compat/xs_lib.h
+# Lib Xen Vchan
+%{_includedir}/libxenvchan.h
+%{_libdir}/libxenvchan.a
+%{_libdir}/libxenvchan.so
+
+%files dom0-tools
+%{_sysconfdir}/bash_completion.d/xl.sh
+%exclude %{_sysconfdir}/rc.d/init.d/xencommons
+%exclude %{_sysconfdir}/rc.d/init.d/xendomains
+%exclude %{_sysconfdir}/rc.d/init.d/xendriverdomain
+%exclude %{_sysconfdir}/sysconfig/xendomains
+%if %with_systemd
+%exclude %{_sysconfdir}/rc.d/init.d/xen-watchdog
+%else
+%{_sysconfdir}/rc.d/init.d/xen-watchdog
+%endif
+%config %{_sysconfdir}/logrotate.d/xen-tools
+%config %{_sysconfdir}/sysconfig/xencommons
+%config %{_sysconfdir}/xen/oxenstored.conf
+%{_sysconfdir}/xen/scripts/block
+%{_sysconfdir}/xen/scripts/block-common.sh
+%{_sysconfdir}/xen/scripts/block-drbd-probe
+%{_sysconfdir}/xen/scripts/block-dummy
+%{_sysconfdir}/xen/scripts/block-enbd
+%{_sysconfdir}/xen/scripts/block-iscsi
+%{_sysconfdir}/xen/scripts/block-nbd
+%{_sysconfdir}/xen/scripts/block-tap
+%{_sysconfdir}/xen/scripts/colo-proxy-setup
+%{_sysconfdir}/xen/scripts/external-device-migrate
+%{_sysconfdir}/xen/scripts/hotplugpath.sh
+%{_sysconfdir}/xen/scripts/locking.sh
+%{_sysconfdir}/xen/scripts/logging.sh
+%{_sysconfdir}/xen/scripts/vif-bridge
+%{_sysconfdir}/xen/scripts/vif-common.sh
+%{_sysconfdir}/xen/scripts/vif-nat
+%{_sysconfdir}/xen/scripts/vif-openvswitch
+%{_sysconfdir}/xen/scripts/vif-route
+%{_sysconfdir}/xen/scripts/vif-setup
+%{_sysconfdir}/xen/scripts/vif2
+%{_sysconfdir}/xen/scripts/vscsi
+%{_sysconfdir}/xen/scripts/xen-hotplug-cleanup
+%{_sysconfdir}/xen/scripts/xen-hotplug-common.sh
+%{_sysconfdir}/xen/scripts/xen-network-common.sh
+%{_sysconfdir}/xen/scripts/xen-script-common.sh
+%exclude %{_sysconfdir}/%{name}/cpupool
+%exclude %{_sysconfdir}/%{name}/README
+%exclude %{_sysconfdir}/%{name}/README.incompatibilities
+%exclude %{_sysconfdir}/%{name}/xlexample.hvm
+%exclude %{_sysconfdir}/%{name}/xlexample.pvlinux
+%config %{_sysconfdir}/xen/xl.conf
+%{_bindir}/pygrub
+%{_bindir}/xen-cpuid
+%{_bindir}/xen-detect
+%{_bindir}/xenalyze
+%{_bindir}/xencons
+%{_bindir}/xencov_split
+%{_bindir}/xentrace_format
+%{python_sitearch}/fsimage.so
+%{python_sitearch}/grub/ExtLinuxConf.py*
+%{python_sitearch}/grub/GrubConf.py*
+%{python_sitearch}/grub/LiloConf.py*
+%{python_sitearch}/grub/__init__.py*
+%{python_sitearch}/pygrub-*.egg-info
+%{python_sitearch}/xen-*.egg-info
+#{python_sitearch}/xen/__init__.py*           - Must not duplicate xen-tools
+#{python_sitearch}/xen/lowlevel/__init__.py*  - Must not duplicate xen-tools
+%{python_sitearch}/xen/lowlevel/xc.so
+%{python_sitearch}/xen/migration/__init__.py*
+%{python_sitearch}/xen/migration/legacy.py*
+%{python_sitearch}/xen/migration/libxc.py*
+%{python_sitearch}/xen/migration/libxl.py*
+%{python_sitearch}/xen/migration/public.py*
+%{python_sitearch}/xen/migration/tests.py*
+%{python_sitearch}/xen/migration/verify.py*
+%{python_sitearch}/xen/migration/xl.py*
+%{_libexecdir}/%{name}/bin/convert-legacy-stream
+%{_libexecdir}/%{name}/bin/init-xenstore-domain
+%{_libexecdir}/%{name}/bin/libxl-save-helper
+%{_libexecdir}/%{name}/bin/lsevtchn
+%{_libexecdir}/%{name}/bin/pygrub
+%{_libexecdir}/%{name}/bin/readnotes
+%{_libexecdir}/%{name}/bin/verify-stream-v2
+%{_libexecdir}/%{name}/bin/xen-init-dom0
+%{_libexecdir}/%{name}/bin/xenconsole
+%{_libexecdir}/%{name}/bin/xenctx
+%{_libexecdir}/%{name}/bin/xendomains
+%{_libexecdir}/%{name}/bin/xenpaging
+%{_libexecdir}/%{name}/bin/xenpvnetboot
+%{_libexecdir}/%{name}/boot/hvmloader
+%{_sbindir}/flask-get-bool
+%{_sbindir}/flask-getenforce
+%{_sbindir}/flask-label-pci
+%{_sbindir}/flask-loadpolicy
+%{_sbindir}/flask-set-bool
+%{_sbindir}/flask-setenforce
+%{_sbindir}/gdbsx
+%{_sbindir}/kdd
+%{_sbindir}/oxenstored
+%{_sbindir}/xen-hptool
+%{_sbindir}/xen-hvmcrash
+%{_sbindir}/xen-hvmctx
+%{_sbindir}/xen-livepatch
+%{_sbindir}/xen-lowmemd
+%{_sbindir}/xen-mceinj
+%{_sbindir}/xen-mfndump
+%exclude %{_sbindir}/xen-ringwatch
+%{_sbindir}/xenbaked
+%{_sbindir}/xenconsoled
+%{_sbindir}/xencov
+%{_sbindir}/xenmon.py
+%{_sbindir}/xenperf
+%{_sbindir}/xenpm
+%{_sbindir}/xenpmd
+%{_sbindir}/xenstored
+%{_sbindir}/xentop
+%{_sbindir}/xentrace
+%{_sbindir}/xentrace_setmask
+%{_sbindir}/xentrace_setsize
+%{_sbindir}/xenwatchdogd
+%{_sbindir}/xl
+%exclude %{_sbindir}/gtracestat
+%exclude %{_sbindir}/gtraceview
+%exclude %{_sbindir}/xen-bugtool
+%exclude %{_sbindir}/xen-tmem-list-parse
+%exclude %{_sbindir}/xenlockprof
+%{_mandir}/man1/xentop.1.gz
+%{_mandir}/man1/xentrace_format.1.gz
+%{_mandir}/man1/xenstore-chmod.1.gz
+%{_mandir}/man1/xenstore-ls.1.gz
+%{_mandir}/man1/xenstore.1.gz
+%{_mandir}/man1/xl.1.gz
+%{_mandir}/man5/xl.cfg.5.gz
+%{_mandir}/man5/xl.conf.5.gz
+%{_mandir}/man5/xlcpupool.cfg.5.gz
+%{_mandir}/man8/xentrace.8.gz
+%dir /var/lib/xen
+%dir /var/log/xen
+%if %with_systemd
+%{_unitdir}/proc-xen.mount
+%{_unitdir}/var-lib-xenstored.mount
+%{_unitdir}/xen-init-dom0.service
+%{_unitdir}/xen-watchdog.service
+%{_unitdir}/xenconsoled.service
+%{_unitdir}/xenstored.service
+%{_unitdir}/xenstored.socket
+%{_unitdir}/xenstored_ro.socket
+%exclude %{_prefix}/lib/modules-load.d/xen.conf
+%exclude %{_unitdir}/xen-qemu-dom0-disk-backend.service
+%exclude %{_unitdir}/xendomains.service
+%endif
+
+%files dom0-libs
+%{_libdir}/fs/ext2fs-lib/fsimage.so
+%{_libdir}/fs/fat/fsimage.so
+%{_libdir}/fs/iso9660/fsimage.so
+%{_libdir}/fs/reiserfs/fsimage.so
+%{_libdir}/fs/ufs/fsimage.so
+%{_libdir}/fs/xfs/fsimage.so
+%{_libdir}/fs/zfs/fsimage.so
+%{_libdir}/libfsimage.so.1.0
+%{_libdir}/libfsimage.so.1.0.0
+%{_libdir}/libxencall.so.1
+%{_libdir}/libxencall.so.1.0
+%{_libdir}/libxenctrl.so.4.7
+%{_libdir}/libxenctrl.so.4.7.0
+%{_libdir}/libxenforeignmemory.so.1
+%{_libdir}/libxenforeignmemory.so.1.0
+%{_libdir}/libxenguest.so.4.7
+%{_libdir}/libxenguest.so.4.7.0
+%{_libdir}/libxenlight.so.4.7
+%{_libdir}/libxenlight.so.4.7.0
+%{_libdir}/libxenstat.so.0
+%{_libdir}/libxenstat.so.0.0
+%{_libdir}/libxentoollog.so.1
+%{_libdir}/libxentoollog.so.1.0
+%{_libdir}/libxlutil.so.4.7
+%{_libdir}/libxlutil.so.4.7.0
+
+%files dom0-libs-devel
+%{_includedir}/fsimage.h
+%{_includedir}/fsimage_grub.h
+%{_includedir}/fsimage_plugin.h
+%{_libdir}/libfsimage.so
+
+%{_includedir}/xencall.h
+%{_libdir}/libxencall.a
+%{_libdir}/libxencall.so
+
+%{_includedir}/xenctrl.h
+%{_includedir}/xenctrl_compat.h
+%{_libdir}/libxenctrl.a
+%{_libdir}/libxenctrl.so
+
+%{_includedir}/xenforeignmemory.h
+%{_libdir}/libxenforeignmemory.a
+%{_libdir}/libxenforeignmemory.so
+
+%{_includedir}/xenguest.h
+%{_libdir}/libxenguest.a
+%{_libdir}/libxenguest.so
+
+%{_includedir}/xentoollog.h
+%{_libdir}/libxentoollog.a
+%{_libdir}/libxentoollog.so
+
+%{_includedir}/_libxl_list.h
+%{_includedir}/_libxl_types.h
+%{_includedir}/_libxl_types_json.h
+%{_includedir}/libxl.h
+%{_includedir}/libxl_event.h
+%{_includedir}/libxl_json.h
+%{_includedir}/libxl_utils.h
+%{_includedir}/libxl_uuid.h
+%{_includedir}/libxlutil.h
+%{_libdir}/libxenlight.a
+%{_libdir}/libxenlight.so
+%{_libdir}/libxlutil.a
+%{_libdir}/libxlutil.so
+/usr/share/pkgconfig/xenlight.pc
+/usr/share/pkgconfig/xlutil.pc
+
+%{_includedir}/xenstat.h
+%{_libdir}/libxenstat.a
+%{_libdir}/libxenstat.so
+
+%files ocaml-libs
+%{_libdir}/ocaml/stublibs/dllxenbus_stubs.so
+%{_libdir}/ocaml/stublibs/dllxenbus_stubs.so.owner
+%{_libdir}/ocaml/stublibs/dllxenctrl_stubs.so
+%{_libdir}/ocaml/stublibs/dllxenctrl_stubs.so.owner
+%{_libdir}/ocaml/stublibs/dllxeneventchn_stubs.so
+%{_libdir}/ocaml/stublibs/dllxeneventchn_stubs.so.owner
+%{_libdir}/ocaml/stublibs/dllxenlight_stubs.so
+%{_libdir}/ocaml/stublibs/dllxenlight_stubs.so.owner
+%{_libdir}/ocaml/stublibs/dllxenmmap_stubs.so
+%{_libdir}/ocaml/stublibs/dllxenmmap_stubs.so.owner
+%{_libdir}/ocaml/stublibs/dllxentoollog_stubs.so
+%{_libdir}/ocaml/stublibs/dllxentoollog_stubs.so.owner
+%{_libdir}/ocaml/xenbus/META
+%{_libdir}/ocaml/xenbus/xenbus.cma
+%{_libdir}/ocaml/xenbus/xenbus.cmo
+%{_libdir}/ocaml/xenctrl/META
+%{_libdir}/ocaml/xenctrl/xenctrl.cma
+%{_libdir}/ocaml/xeneventchn/META
+%{_libdir}/ocaml/xeneventchn/xeneventchn.cma
+%{_libdir}/ocaml/xenlight/META
+%{_libdir}/ocaml/xenlight/xenlight.cma
+%{_libdir}/ocaml/xenmmap/META
+%{_libdir}/ocaml/xenmmap/xenmmap.cma
+%exclude %{_libdir}/ocaml/xenstore/META
+%exclude %{_libdir}/ocaml/xenstore/xenstore.cma
+%exclude %{_libdir}/ocaml/xenstore/xenstore.cmo
+%{_libdir}/ocaml/xentoollog/META
+%{_libdir}/ocaml/xentoollog/xentoollog.cma
+
+%files ocaml-devel
+%{_libdir}/ocaml/xenbus/libxenbus_stubs.a
+%{_libdir}/ocaml/xenbus/xenbus.a
+%{_libdir}/ocaml/xenbus/xenbus.cmi
+%{_libdir}/ocaml/xenbus/xenbus.cmx
+%{_libdir}/ocaml/xenbus/xenbus.cmxa
+%{_libdir}/ocaml/xenctrl/libxenctrl_stubs.a
+%{_libdir}/ocaml/xenctrl/xenctrl.a
+%{_libdir}/ocaml/xenctrl/xenctrl.cmi
+%{_libdir}/ocaml/xenctrl/xenctrl.cmx
+%{_libdir}/ocaml/xenctrl/xenctrl.cmxa
+%{_libdir}/ocaml/xeneventchn/libxeneventchn_stubs.a
+%{_libdir}/ocaml/xeneventchn/xeneventchn.a
+%{_libdir}/ocaml/xeneventchn/xeneventchn.cmi
+%{_libdir}/ocaml/xeneventchn/xeneventchn.cmx
+%{_libdir}/ocaml/xeneventchn/xeneventchn.cmxa
+%{_libdir}/ocaml/xenlight/libxenlight_stubs.a
+%{_libdir}/ocaml/xenlight/xenlight.a
+%{_libdir}/ocaml/xenlight/xenlight.cmi
+%{_libdir}/ocaml/xenlight/xenlight.cmx
+%{_libdir}/ocaml/xenlight/xenlight.cmxa
+%{_libdir}/ocaml/xenmmap/libxenmmap_stubs.a
+%{_libdir}/ocaml/xenmmap/xenmmap.a
+%{_libdir}/ocaml/xenmmap/xenmmap.cmi
+%{_libdir}/ocaml/xenmmap/xenmmap.cmx
+%{_libdir}/ocaml/xenmmap/xenmmap.cmxa
+%exclude %{_libdir}/ocaml/xenstore/xenstore.a
+%exclude %{_libdir}/ocaml/xenstore/xenstore.cmi
+%exclude %{_libdir}/ocaml/xenstore/xenstore.cmx
+%exclude %{_libdir}/ocaml/xenstore/xenstore.cmxa
+%{_libdir}/ocaml/xentoollog/libxentoollog_stubs.a
+%{_libdir}/ocaml/xentoollog/xentoollog.a
+%{_libdir}/ocaml/xentoollog/xentoollog.cmi
+%{_libdir}/ocaml/xentoollog/xentoollog.cmx
+%{_libdir}/ocaml/xentoollog/xentoollog.cmxa
+
+%files installer-files
+%{_libdir}/libxenctrl.so.4.7
+%{_libdir}/libxenctrl.so.4.7.0
+%{_libdir}/libxenguest.so.4.7
+%{_libdir}/libxenguest.so.4.7.0
+%{python_sitearch}/xen/__init__.py*
+%{python_sitearch}/xen/lowlevel/__init__.py*
+%{python_sitearch}/xen/lowlevel/xc.so
+
+%doc
+
+%post hypervisor
+# Update the debug and release symlinks
+ln -sf %{name}-%{version}%{vendor_version}-d.gz /boot/xen-debug.gz
+ln -sf %{name}-%{version}%{vendor_version}.gz /boot/xen-release.gz
+
+# Point /boot/xen.gz appropriately
+if [ ! -e /boot/xen.gz ]; then
+    # Use a debug hypervisor by default
+    ln -sf %{name}-%{version}%{vendor_version}-d.gz /boot/xen.gz
+else
+    # Else look at the current link, and whether it is debug
+    path="`readlink -f /boot/xen.gz`"
+    if [ ${path} != ${path%%-d.gz} ]; then
+        ln -sf %{name}-%{version}%{vendor_version}-d.gz /boot/xen.gz
+    else
+        ln -sf %{name}-%{version}%{vendor_version}.gz /boot/xen.gz
+    fi
+fi
+
+if [ -e %{_sysconfdir}/sysconfig/kernel ] && ! grep -q '^HYPERVISOR' %{_sysconfdir}/sysconfig/kernel ; then
+  cat %{_sysconfdir}/sysconfig/kernel-xen >> %{_sysconfdir}/sysconfig/kernel
+fi
+
+mkdir -p %{_rundir}/reboot-required.d/%{name}
+touch %{_rundir}/reboot-required.d/%{name}/%{version}-%{release}
+
+%if %with_systemd
+%post dom0-tools
+%systemd_post proc-xen.mount
+%systemd_post var-lib-xenstored.mount
+%systemd_post xen-init-dom0.service
+%systemd_post xen-watchdog.service
+%systemd_post xenconsoled.service
+%systemd_post xenstored.service
+%systemd_post xenstored.socket
+%systemd_post xenstored_ro.socket
+
+%preun dom0-tools
+%systemd_preun proc-xen.mount
+%systemd_preun var-lib-xenstored.mount
+%systemd_preun xen-init-dom0.service
+%systemd_preun xen-watchdog.service
+%systemd_preun xenconsoled.service
+%systemd_preun xenstored.service
+%systemd_preun xenstored.socket
+%systemd_preun xenstored_ro.socket
+
+%postun dom0-tools
+%systemd_postun proc-xen.mount
+%systemd_postun var-lib-xenstored.mount
+%systemd_postun xen-init-dom0.service
+%systemd_postun xen-watchdog.service
+%systemd_postun xenconsoled.service
+%systemd_postun xenstored.service
+%systemd_postun xenstored.socket
+%systemd_postun xenstored_ro.socket
+%endif
+
diff --git a/mk/xl.conf b/mk/xl.conf
new file mode 100644
index 0000000..3da0a1e
--- /dev/null
+++ b/mk/xl.conf
@@ -0,0 +1 @@
+autoballoon = 0
