# HG changeset patch
# Parent 1ad1537e8bb543aafd9f41afe5d5b475a34c59fa

diff --git a/tools/firmware/hvmloader/config.h b/tools/firmware/hvmloader/config.h
index b6f3d33..72ed0be 100644
--- a/tools/firmware/hvmloader/config.h
+++ b/tools/firmware/hvmloader/config.h
@@ -3,7 +3,7 @@
 
 #include <stdint.h>
 
-enum virtual_vga { VGA_none, VGA_std, VGA_cirrus, VGA_pt };
+enum virtual_vga { VGA_none, VGA_std, VGA_cirrus, VGA_pt, VGA_vgt };
 extern enum virtual_vga virtual_vga;
 
 extern unsigned long igd_opregion_addr;
diff --git a/tools/firmware/hvmloader/pci.c b/tools/firmware/hvmloader/pci.c
index 24b068d..85c9153 100644
--- a/tools/firmware/hvmloader/pci.c
+++ b/tools/firmware/hvmloader/pci.c
@@ -23,6 +23,7 @@
 #include "hypercall.h"
 #include "config.h"
 #include "pci_regs.h"
+#include "vgt.h"
 
 #include <xen/memory.h>
 #include <xen/hvm/ioreq.h>
@@ -84,6 +85,7 @@ void pci_setup(void)
     uint32_t vga_devfn = 256;
     uint16_t class, vendor_id, device_id;
     unsigned int bar, pin, link, isa_irq;
+    uint8_t intel_gpu_present = 0;
 
     /* Resources assignable to PCI devices via BARs. */
     struct resource {
@@ -158,6 +160,7 @@ void pci_setup(void)
 
         if (( vendor_id == 0x8086 ) && (class == 0x0300 || class == 0x0380))
         {
+            intel_gpu_present = 1;
             igd_opregion_addr = mem_hole_alloc(IGD_OPREGION_PAGES);
             /*
              * Write the the OpRegion offset to give the opregion
@@ -531,6 +534,36 @@ void pci_setup(void)
 
     if ( vga_devfn != 256 )
     {
+        if ( intel_gpu_present )
+        {
+            uint32_t bar = pci_readl(vga_devfn, PCI_BASE_ADDRESS_0)
+                                        & PCI_BASE_ADDRESS_MEM_MASK;
+
+            void *pvinfo = (void *)bar + VGT_PVINFO_PAGE;
+            uint64_t *magic = pvinfo;
+
+            if (*magic == VGT_MAGIC) {
+		uint32_t vga_bar = pci_readl(vga_devfn, PCI_BASE_ADDRESS_5)
+						& PCI_BASE_ADDRESS_MEM_MASK;
+                /*
+                 * Found VGT device, and use standard VGA bios.
+                 */
+                virtual_vga = VGA_vgt;
+
+                /* XXX: we use this hack to tell vGT driver the
+                 * top of <4G mem, so vGT can avoid unnecessary
+                 * attempts to map the mem hole. This optimization
+                 * can speed up guest bootup time and improve Win7
+                 * SMP guest's stability.
+                 * NOTE: here we're actually trying to write 32 bits
+                 * into VENDOR_ID and DEVICE_ID -- we assume normally
+                 * sane codes in guest won't do this...
+                 */
+                 pci_writel(vga_devfn, PCI_VENDOR_ID, hvm_info->low_mem_pgend);
+		 /* XXX: Let QEMU legacy VGA emulator know the vga bar address*/
+                 pci_writel(0, 0xf8, vga_bar);
+            }
+        }
         /*
          * VGA registers live in I/O space so ensure that primary VGA
          * has IO enabled, even if there is no I/O BAR on that
diff --git a/tools/firmware/hvmloader/rombios.c b/tools/firmware/hvmloader/rombios.c
index 4e9eff1..fbdc461 100644
--- a/tools/firmware/hvmloader/rombios.c
+++ b/tools/firmware/hvmloader/rombios.c
@@ -92,6 +92,7 @@ static void rombios_load_roms(void)
                vgabios_cirrusvga, sizeof(vgabios_cirrusvga));
         vgabios_sz = round_option_rom(sizeof(vgabios_cirrusvga));
         break;
+    case VGA_vgt:
     case VGA_std:
         printf("Loading Standard VGABIOS ...\n");
         memcpy((void *)VGABIOS_PHYSICAL_ADDRESS,
diff --git a/tools/firmware/hvmloader/vgt.h b/tools/firmware/hvmloader/vgt.h
new file mode 100644
index 0000000..f5b3f34
--- /dev/null
+++ b/tools/firmware/hvmloader/vgt.h
@@ -0,0 +1,24 @@
+/*
+ * Copyright (c) 2012-2013, Intel Corporation.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef _VGT_DEVTABLE_H
+#define _VGT_DEVTABLE_H
+
+#define VGT_PVINFO_PAGE 	0x78000
+#define VGT_MAGIC		0x4776544776544776    /* 'vGTvGTvG' */
+
+#endif  /* _VGT_DEVTABLE_H */
