diff --git a/tools/ocaml/libs/xb/partial.ml b/tools/ocaml/libs/xb/partial.ml
index d4d1c7b..025a5d7 100644
--- a/tools/ocaml/libs/xb/partial.ml
+++ b/tools/ocaml/libs/xb/partial.ml
@@ -27,6 +27,8 @@ external header_size: unit -> int = "stub_header_size"
 external header_of_string_internal: string -> int * int * int * int
          = "stub_header_of_string"
 
+let allow_oversize_packets = ref true
+
 let xenstore_payload_max = 4096 (* xen/include/public/io/xs_wire.h *)
 
 let of_string s =
@@ -35,7 +37,7 @@ let of_string s =
 	   This will leave the guest connection is a bad state and will
 	   be hard to recover from without restarting the connection
 	   (ie rebooting the guest) *)
-	let dlen = min xenstore_payload_max dlen in
+	let dlen = if !allow_oversize_packets then dlen else min xenstore_payload_max dlen in
 	{
 		tid = tid;
 		rid = rid;
@@ -45,7 +47,7 @@ let of_string s =
 	}
 
 let append pkt s sz =
-	if pkt.len > 4096 then failwith "Buffer.add: cannot grow buffer";
+	if not !allow_oversize_packets && pkt.len > 4096 then failwith "Buffer.add: cannot grow buffer";
 	Buffer.add_string pkt.buf (String.sub s 0 sz)
 
 let to_complete pkt =
diff --git a/tools/ocaml/xenstored/xenstored.ml b/tools/ocaml/xenstored/xenstored.ml
index 1c02f2f..0bb34a7 100644
--- a/tools/ocaml/xenstored/xenstored.ml
+++ b/tools/ocaml/xenstored/xenstored.ml
@@ -212,6 +212,8 @@ let to_file store cons file =
 end
 
 let _ =
+	Xenbus.Partial.allow_oversize_packets := false;
+
 	let cf = do_argv in
 	let pidfile =
 		if Sys.file_exists (config_filename cf) then
