diff --git a/tools/pygrub/src/pygrub b/tools/pygrub/src/pygrub
index ae2f21d..8a9734a 100644
--- a/tools/pygrub/src/pygrub
+++ b/tools/pygrub/src/pygrub
@@ -885,7 +885,15 @@ if __name__ == "__main__":
     if len(part_offs) < 1:
         raise RuntimeError, "Disk has no partitions"
 
-    fs = fsimage.open(file, part_offs[0], bootfsoptions)
+    try:
+        fs = fsimage.open(file, part_offs[0], bootfsoptions)
+    except IOError:
+        # in this case we tried to detect a disk with a direct filesystem
+        # but probably this disk does not have a valid partition
+        if len(part_offs) == 1 and part_offs[0] == 0:
+            raise RuntimeError, "Bootable disk has no valid filesystem"
+        else:
+            raise
 
     # We always boot the "default" kernel if it exists, rather than
     # parsing the grub menu
