diff --git a/xen/drivers/passthrough/amd/iommu_init.c b/xen/drivers/passthrough/amd/iommu_init.c
index 56bda00..91dbf51 100644
--- a/xen/drivers/passthrough/amd/iommu_init.c
+++ b/xen/drivers/passthrough/amd/iommu_init.c
@@ -1215,7 +1215,7 @@ static bool_t __init amd_sp5100_erratum28(void)
         byte = pci_conf_read8(0, bus, 0x14, 0, 0xad);
         if ( (byte >> 3) & 1 )
         {
-            printk(XENLOG_WARNING "AMD-Vi: SP5100 erratum 28 detected, disabling IOMMU.\n"
+            printk(XENLOG_WARNING "AMD-Vi: SP5100 erratum 28 detected, disabling Interrupt Remapping.\n"
                    "If possible, disable SATA Combined mode in BIOS or contact your vendor for BIOS update.\n");
             return 1;
         }
@@ -1232,7 +1232,9 @@ int __init amd_iommu_init(void)
 
     if ( iommu_intremap && amd_iommu_perdev_intremap &&
          amd_sp5100_erratum28() )
-        goto error_out;
+    {
+        iommu_intremap = 0;
+    }
 
     ivrs_bdf_entries = amd_iommu_get_ivrs_dev_entries();
 
diff --git a/xen/drivers/passthrough/vtd/quirks.c b/xen/drivers/passthrough/vtd/quirks.c
index 839dc45..9415d46 100644
--- a/xen/drivers/passthrough/vtd/quirks.c
+++ b/xen/drivers/passthrough/vtd/quirks.c
@@ -253,8 +253,8 @@ static void __init tylersburg_intremap_quirk(void)
         if ( rev == 0x13 && device == 0x342e8086 )
         {
             printk(XENLOG_WARNING VTDPREFIX
-                   "Disabling IOMMU due to Intel 5500/5520/X58 Chipset errata #47, #53\n");
-            iommu_enable = 0;
+                   "Disabling Interrupt remapping due to Intel 5500/5520/X58 Chipset errata #47, #53\n");
+            iommu_intremap = 0;
             break;
         }
     }
